### PR TITLE
Nothing is impossible

### PR DESCRIPTION
**Pull request by ZeroDeth**
‘Nothing is impossible – the word itself says, “I’m possible!”’ Audrey Hepburn

### CODE REVIEW

**Comment by ZeroDeth** [comment]
My plan was to move Struct under every function. TBH, IDK is a good or bad thing.

**Comment by ZeroDeth** [comment]
Exercise 7. Root music, a little bit harder, probably should recap and try testing more my skills with exercism.io then come back with new PR for Q7. Thanks

File: calculator.go [hunk 1]
```diff
@@ -1,13 +1,28 @@
...
+
+// Multiply takes two numbers as parameters, and returns a single number representing the result.
+func Multiply(a, b float64) float64 {
+	return a * b
+}
+
+// Divide takes two numbers as parameters, and returns first divide by second representing the result.
+func Divide(a, b float64) (float64, error) {
+	if b == 0 {
+		return 0, fmt.Errorf("division by zero is undefined")
```
Comment (bitfield) [inline calculator.go]: If we're not interpolating any data, then we don't need `fmt.Errorf`—we can use `errors.New` instead.

File: calculator_test.go [hunk 2]
```diff
@@ -2,9 +2,18 @@ package calculator_test
...
 	"calculator"
+	"math/rand"
 	"testing"
+	"time"
 )
 
+type testCase struct {
+	a, b        float64
+	want        float64
+	name        string
```
Comment (bitfield) [inline calculator_test.go]: We might put the name first, to make the cases a bit more readable, mightn't we?

File: calculator_test.go [hunk 3]
```diff
@@ -14,11 +23,90 @@ func TestAdd(t *testing.T) {
...
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func TestAdd1and1(t *testing.T) {
```
Comment (bitfield) [inline calculator_test.go]: We no longer really need this separate test if it's covered by TestAddStruct, do we?

File: calculator_test.go [hunk 4]
```diff
@@ -14,11 +23,90 @@ func TestAdd(t *testing.T) {
...
+		{a: -1, b: -1, want: -2, name: "Two negative numbers that sum ot a negative"},
+		{a: -5, b: 0, want: -5, name: "Negative and positive number that sum Zero"},
+	}
+
+	// Looping over test cases
+	for _, tc := range testCases {
+		got := calculator.Add(tc.a, tc.b)
+		if tc.want != got {
+			t.Errorf("%s", tc.name)
+			t.Errorf("Add(%f, %f): want %f, got %f", tc.a, tc.b, tc.want, got) // Todo: Why cannot add (tc.name) here
```
Comment (bitfield) [inline calculator_test.go]:
```suggestion
			t.Errorf("%s: Add(%f, %f): want %f, got %f", tc.name, tc.a, tc.b, tc.want, got)
```

File: calculator_test.go [hunk 5]
```diff
@@ -14,11 +23,90 @@ func TestAdd(t *testing.T) {
...
+		got := calculator.Add(a, b)
+		if sum != got {
+			t.Fatalf("Add(%f, %f) ; want %f, got %f", a, b, sum, got)
+		}
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	t.Parallel()
+	var want float64 = 2
```
Comment (bitfield) [inline calculator_test.go]: Try adding a slice of test cases for TestSubtract and TestMultiply too!

File: calculator_test.go [hunk 6]
```diff
@@ -14,11 +23,90 @@ func TestAdd(t *testing.T) {
...
+	for _, tc := range testCases {
+		got, err := calculator.Divide(tc.a, tc.b)
+		errReceived := err != nil
+		if tc.errExpected != errReceived {
+			t.Fatalf("Divide(%f, %f): unexpected error status: %v", tc.a, tc.b, errReceived)
+
+			// A shorter way to write this logic is:
+			/*
+			    if tc.errExpected != (err != nil) {
+			    t.Fatalf(...) // Todo: Not working
```
Comment (bitfield) [inline calculator_test.go]: I'm not sure why this isn't working for you! Try it again.

**Review by bitfield — COMMENTED** [review]
Looking very nice!

**Comment by ZeroDeth** [comment]
> If we're not interpolating any data, then we don't need `fmt.Errorf`—we can use `errors.New` instead. 

Investigating more but changed. Thx

**Comment by ZeroDeth** [comment]
> I'm not sure why this isn't working for you! Try it again. 

Not enough arguments in call 't.Fatalf' - to expected operand, found '...'

File: calculator_test.go (hunk 6 above)
Comment (ZeroDeth) [reply calculator_test.go]: Not enough arguments in call 't.Fatalf' - to expected operand, found '...'

Comment (ZeroDeth) [reply calculator_test.go]: CleanUp. Thx

**Pull request merged**